{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Importing necessary packages\n",
    "import cobra\n",
    "import itertools\n",
    "import metquest\n",
    "import networkx as nx\n",
    "import orthoani\n",
    "import os\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import re\n",
    "import shutil\n",
    "import sys\n",
    "import tempfile\n",
    "from Bio.SeqIO import parse\n",
    "from cobra.medium import minimal_medium\n",
    "from itertools import combinations\n",
    "from libsbml import SBMLDocument, SBMLReader\n",
    "from metquest import create_graph\n",
    "from typing import Dict, List, Set, Tuple, Union\n",
    "\n",
    "# Loading necessary path\n",
    "input_path = \"/mnt/Local_Disk_1/3_MiMB_Book_Chapter/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2024-12-26 17:03:14--  http://bigg.ucsd.edu/static/models/iCN718.xml\n",
      "Resolving bigg.ucsd.edu (bigg.ucsd.edu)... 169.228.33.117\n",
      "Connecting to bigg.ucsd.edu (bigg.ucsd.edu)|169.228.33.117|:80... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 4921390 (4.7M) [application/xml]\n",
      "Saving to: ‘/mnt/Local_Disk_1/3_MiMB_Book_Chapter/1_Models/iCN718.xml’\n",
      "\n",
      "iCN718.xml          100%[===================>]   4.69M  1.90MB/s    in 2.5s    \n",
      "\n",
      "2024-12-26 17:03:18 (1.90 MB/s) - ‘/mnt/Local_Disk_1/3_MiMB_Book_Chapter/1_Models/iCN718.xml’ saved [4921390/4921390]\n",
      "\n",
      "--2024-12-26 17:03:19--  http://bigg.ucsd.edu/static/models/iYL1228.xml\n",
      "Resolving bigg.ucsd.edu (bigg.ucsd.edu)... 169.228.33.117\n",
      "Connecting to bigg.ucsd.edu (bigg.ucsd.edu)|169.228.33.117|:80... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 9202114 (8.8M) [application/xml]\n",
      "Saving to: ‘/mnt/Local_Disk_1/3_MiMB_Book_Chapter/1_Models/iYL1228.xml’\n",
      "\n",
      "iYL1228.xml         100%[===================>]   8.78M  2.69MB/s    in 3.3s    \n",
      "\n",
      "2024-12-26 17:03:22 (2.69 MB/s) - ‘/mnt/Local_Disk_1/3_MiMB_Book_Chapter/1_Models/iYL1228.xml’ saved [9202114/9202114]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Collecting all the SBML files\n",
    "# Collecting the Acinetobacter baumannii AYE (iCN718) and\n",
    "# Klebsiella pneumoniae subsp. pneumoniae MGH 78578 (iYL1228) models\n",
    "\n",
    "!wget http://bigg.ucsd.edu/static/models/iCN718.xml -P \"/mnt/Local_Disk_1/3_MiMB_Book_Chapter/1_Models/\"\n",
    "!wget http://bigg.ucsd.edu/static/models/iYL1228.xml -P \"/mnt/Local_Disk_1/3_MiMB_Book_Chapter/1_Models/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "99.80144101731602\n",
      "99.0398276003276\n"
     ]
    }
   ],
   "source": [
    "# Calculating pairwise average nucleotide identity (ANI)\n",
    "# Between Acinetobacter baumannii AYE (GCF_000069245.1) and \n",
    "# and AB6870155 (GCF_000794125.2)\n",
    "!orthoani -q /mnt/Local_Disk_1/3_MiMB_Book_Chapter/2_Genomes/GCF_000069245.1.fna -r /mnt/Local_Disk_1/3_MiMB_Book_Chapter/2_Genomes/GCF_000794125.2.fna -j 12\n",
    "\n",
    "# Calculating pairwise average nucleotide identity (ANI)\n",
    "# Between Klebsiella pneumoniae subsp. pneumoniae MGH 78578 \n",
    "# (GCF_000016305.1) and KP6870155 (GCF_027126715.1)\n",
    "!orthoani -q /mnt/Local_Disk_1/3_MiMB_Book_Chapter/2_Genomes/GCF_000016305.1.fna -r /mnt/Local_Disk_1/3_MiMB_Book_Chapter/2_Genomes/GCF_027126715.1.fna -j 12"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The current solver interface glpk doesn't support setting the optimality tolerance.\n",
      "\u001b[1m============================= test session starts ==============================\u001b[0m\n",
      "platform linux -- Python 3.10.16, pytest-8.3.4, pluggy-1.5.0\n",
      "rootdir: /home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests\n",
      "plugins: typeguard-4.4.1, anyio-4.7.0\n",
      "collected 146 items / 1 skipped                                                \u001b[0m\n",
      "\n",
      "../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py \u001b[32m.\u001b[0m\u001b[33m [  0%]\n",
      "\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[31m         [ 44%]\u001b[0m\n",
      "../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_basic.py \u001b[32m.\u001b[0m\u001b[31m [ 45%]\n",
      "\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31m                                                   [ 60%]\u001b[0m\n",
      "../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_biomass.py \u001b[32m.\u001b[0m\u001b[31m [ 60%]\n",
      "\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[31m                                                                [ 67%]\u001b[0m\n",
      "../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py \u001b[31mF\u001b[0m\u001b[31m [ 67%]\n",
      "\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[33ms\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[33ms\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31m                                             [ 86%]\u001b[0m\n",
      "../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_essentiality.py \u001b[33ms\u001b[0m\u001b[31m [ 87%]\n",
      "\u001b[0m\u001b[31m                                                                         [ 87%]\u001b[0m\n",
      "../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_growth.py \u001b[33ms\u001b[0m\u001b[31m [ 88%]\n",
      "\u001b[0m\u001b[31m                                                                         [ 88%]\u001b[0m\n",
      "../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_matrix.py \u001b[32m.\u001b[0m\u001b[31m [ 89%]\n",
      "\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[31m                                                                      [ 91%]\u001b[0m\n",
      "../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_sbml.py \u001b[32m.\u001b[0m\u001b[31m [ 91%]\n",
      "\u001b[0m\u001b[32m.\u001b[0m\u001b[31m                                                                        [ 92%]\u001b[0m\n",
      "../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_sbo.py \u001b[31mF\u001b[0m\u001b[31m [ 93%]\n",
      "\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[31m                                                               [100%]\u001b[0m\n",
      "\n",
      "\u001b[33m=============================== warnings summary ===============================\u001b[0m\n",
      "../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/_pytest/config/__init__.py:1277\n",
      "  /home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/_pytest/config/__init__.py:1277: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: typeguard\n",
      "    self._mark_plugins_for_rewrite(hook)\n",
      "\n",
      "test_annotation.py::test_metabolite_id_namespace_consistency\n",
      "test_annotation.py::test_reaction_id_namespace_consistency\n",
      "  /home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/support/annotation.py:279: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.\n",
      "  The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.\n",
      "  \n",
      "  For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.\n",
      "  \n",
      "  \n",
      "    df[\"duplicate\"].fillna(False, inplace=True)\n",
      "\n",
      "test_annotation.py::test_metabolite_id_namespace_consistency\n",
      "  /home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/support/annotation.py:279: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "    df[\"duplicate\"].fillna(False, inplace=True)\n",
      "\n",
      "-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html\n",
      "\u001b[36m\u001b[1m=========================== short test summary info ============================\u001b[0m\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_overview[pubchem.compound]\u001b[0m - AssertionError: The following 888 metabolites (100.00%) lack annotation for\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_overview[kegg.compound]\u001b[0m - AssertionError: The following 142 metabolites (15.99%) lack annotation for\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_overview[seed.compound]\u001b[0m - AssertionError: The following 119 metabolites (13.40%) lack annotation for\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_overview[inchikey]\u001b[0m - AssertionError: The following 888 metabolites (100.00%) lack annotation for...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_overview[inchi]\u001b[0m - AssertionError: The following 888 metabolites (100.00%) lack annotation for...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_overview[chebi]\u001b[0m - AssertionError: The following 118 metabolites (13.29%) lack annotation for ...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_overview[hmdb]\u001b[0m - AssertionError: The following 293 metabolites (33.00%) lack annotation for ...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_overview[reactome]\u001b[0m - AssertionError: The following 888 metabolites (100.00%) lack annotation for...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_overview[metanetx.chemical]\u001b[0m - AssertionError: The following 78 metabolites (8.78%) lack annotation for\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_overview[biocyc]\u001b[0m - AssertionError: The following 150 metabolites (16.89%) lack annotation for ...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_reaction_annotation_overview[rhea]\u001b[0m - AssertionError: The following 445 reactions (43.84%) lack annotation for rhea:\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_reaction_annotation_overview[kegg.reaction]\u001b[0m - AssertionError: The following 517 reactions (50.94%) lack annotation for\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_reaction_annotation_overview[seed.reaction]\u001b[0m - AssertionError: The following 280 reactions (27.59%) lack annotation for\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_reaction_annotation_overview[metanetx.reaction]\u001b[0m - AssertionError: The following 177 reactions (17.44%) lack annotation for\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_reaction_annotation_overview[reactome]\u001b[0m - AssertionError: The following 1015 reactions (100.00%) lack annotation for ...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_reaction_annotation_overview[ec-code]\u001b[0m - AssertionError: The following 433 reactions (42.66%) lack annotation for ec...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_reaction_annotation_overview[brenda]\u001b[0m - AssertionError: The following 1015 reactions (100.00%) lack annotation for ...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_reaction_annotation_overview[biocyc]\u001b[0m - AssertionError: The following 472 reactions (46.50%) lack annotation for bi...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_overview[refseq]\u001b[0m - AssertionError: The following 709 genes (100.00%) lack annotation for refseq:\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_overview[uniprot]\u001b[0m - AssertionError: The following 709 genes (100.00%) lack annotation for uniprot:\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_overview[ecogene]\u001b[0m - AssertionError: The following 709 genes (100.00%) lack annotation for ecogene:\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_overview[kegg.genes]\u001b[0m - AssertionError: The following 709 genes (100.00%) lack annotation for kegg....\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_overview[ncbigi]\u001b[0m - AssertionError: The following 709 genes (100.00%) lack annotation for ncbigi:\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_overview[ncbigene]\u001b[0m - AssertionError: The following 709 genes (100.00%) lack annotation for ncbig...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_overview[ncbiprotein]\u001b[0m - AssertionError: The following 709 genes (100.00%) lack annotation for ncbip...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_overview[ccds]\u001b[0m - AssertionError: The following 709 genes (100.00%) lack annotation for ccds:\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_overview[hprd]\u001b[0m - AssertionError: The following 709 genes (100.00%) lack annotation for hprd:\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_overview[asap]\u001b[0m - AssertionError: The following 709 genes (100.00%) lack annotation for asap:\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_wrong_ids[pubchem.compound]\u001b[0m - AssertionError: There are no metabolite annotations for the pubchem.compoun...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_wrong_ids[inchikey]\u001b[0m - AssertionError: There are no metabolite annotations for the inchikey database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_wrong_ids[inchi]\u001b[0m - AssertionError: There are no metabolite annotations for the inchi database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_wrong_ids[reactome]\u001b[0m - AssertionError: There are no metabolite annotations for the reactome database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_reaction_annotation_wrong_ids[rhea]\u001b[0m - AssertionError: A total of 21 reaction annotations (2.07%) do not match the\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_reaction_annotation_wrong_ids[reactome]\u001b[0m - AssertionError: There are no reaction annotations for the reactome database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_reaction_annotation_wrong_ids[ec-code]\u001b[0m - AssertionError: A total of 22 reaction annotations (2.17%) do not match the\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_reaction_annotation_wrong_ids[brenda]\u001b[0m - AssertionError: There are no reaction annotations for the brenda database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_wrong_ids[refseq]\u001b[0m - AssertionError: There are no gene annotations for the refseq database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_wrong_ids[uniprot]\u001b[0m - AssertionError: There are no gene annotations for the uniprot database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_wrong_ids[ecogene]\u001b[0m - AssertionError: There are no gene annotations for the ecogene database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_wrong_ids[kegg.genes]\u001b[0m - AssertionError: There are no gene annotations for the kegg.genes database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_wrong_ids[ncbigi]\u001b[0m - AssertionError: There are no gene annotations for the ncbigi database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_wrong_ids[ncbigene]\u001b[0m - AssertionError: There are no gene annotations for the ncbigene database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_wrong_ids[ncbiprotein]\u001b[0m - AssertionError: There are no gene annotations for the ncbiprotein database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_wrong_ids[ccds]\u001b[0m - AssertionError: There are no gene annotations for the ccds database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_wrong_ids[hprd]\u001b[0m - AssertionError: There are no gene annotations for the hprd database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_wrong_ids[asap]\u001b[0m - AssertionError: There are no gene annotations for the asap database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_basic.py::\u001b[1mtest_metabolites_formula_presence\u001b[0m - AssertionError: There are a total of 720         metabolites (81.08%) witho...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_basic.py::\u001b[1mtest_gene_protein_reaction_rule_presence\u001b[0m - AssertionError: There are a total of 150 reactions (14.78%) without GPR:\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_basic.py::\u001b[1mtest_transport_reaction_gpr_presence\u001b[0m - AssertionError: There are a total of 84 transport reactions (67.20% of all\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_basic.py::\u001b[1mtest_find_reactions_with_partially_identical_annotations\u001b[0m - AssertionError: Based on annotations there are 25 different groups of overl...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_basic.py::\u001b[1mtest_find_reactions_with_identical_genes\u001b[0m - AssertionError: Based only on equal genes there are 133 different groups of\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_basic.py::\u001b[1mtest_find_medium_metabolites\u001b[0m - TypeError: unsupported operand type(s) for /: 'int' and 'list'\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_biomass.py::\u001b[1mtest_biomass_consistency[BIOMASS__3]\u001b[0m - AssertionError: The component molar mass of the biomass reaction BIOMASS__3...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_biomass.py::\u001b[1mtest_gam_in_biomass[BIOMASS__3]\u001b[0m - AssertionError: No, BIOMASS__3 does not contain a term for growth-associated\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_biomass.py::\u001b[1mtest_fast_growth_default[BIOMASS__3]\u001b[0m - AssertionError: Using the biomass reaction BIOMASS__3 and when the model is...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_biomass.py::\u001b[1mtest_essential_precursors_not_in_biomass[BIOMASS__3]\u001b[0m - AssertionError: BIOMASS__3 lacks a total of 8 essential metabolites        ...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_stoichiometric_consistency\u001b[0m - AssertionError: This model's stoichiometry inconsistent\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_unconserved_metabolites\u001b[0m - AssertionError: This model contains 600 unconserved metabolites: dopa_c, da...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_inconsistent_min_stoichiometry\u001b[0m - TypeError: sequence item 0: expected str instance, list found\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_detect_energy_generating_cycles[MNXM6]\u001b[0m - AssertionError: The model can produce 'MNXM6' without requiring resources. ...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_detect_energy_generating_cycles[MNXM10]\u001b[0m - AssertionError: The model can produce 'MNXM10' without requiring resources....\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_detect_energy_generating_cycles[MNXM38]\u001b[0m - AssertionError: The model can produce 'MNXM38' without requiring resources....\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_detect_energy_generating_cycles[MNXM191]\u001b[0m - AssertionError: The model can produce 'MNXM191' without requiring resources...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_detect_energy_generating_cycles[MNXM223]\u001b[0m - AssertionError: The model can produce 'MNXM223' without requiring resources...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_detect_energy_generating_cycles[MNXM12233]\u001b[0m - AssertionError: The model can produce 'MNXM12233' without requiring resourc...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_detect_energy_generating_cycles[MNXM558]\u001b[0m - AssertionError: The model can produce 'MNXM558' without requiring resources...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_detect_energy_generating_cycles[MNXM89557]\u001b[0m - RuntimeError: It was not possible to uniquely identify a single metabolite ...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_reaction_charge_balance\u001b[0m - AssertionError: A total of 32 (3.61%) reactions are charge unbalanced with at\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_reaction_mass_balance\u001b[0m - AssertionError: A total of 882 (99.55%) reactions are mass unbalanced with ...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_blocked_reactions\u001b[0m - AssertionError: There are 392 (38.62%) blocked reactions in         the mod...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_find_stoichiometrically_balanced_cycles\u001b[0m - AssertionError: There are 79 (7.78%) reactions         which participate in...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_find_orphans\u001b[0m - AssertionError: A total of 69 (7.77%) metabolites are not produced by any r...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_find_deadends\u001b[0m - AssertionError: A total of 92 (10.36%) metabolites are not consumed by any ...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_find_disconnected\u001b[0m - AssertionError: A total of 4 (0.45%) metabolites are not associated with any\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_find_metabolites_not_produced_with_open_bounds\u001b[0m - AssertionError: A total of 252 (28.38%) metabolites cannot be produced in c...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_find_metabolites_not_consumed_with_open_bounds\u001b[0m - AssertionError: A total of 346 (38.96%) metabolites cannot be consumed in c...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_find_reactions_unbounded_flux_default_condition\u001b[0m - AssertionError:  A fraction of 18.30% of the non-blocked reactions (in tota...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_sbo.py::\u001b[1mtest_metabolite_sbo_presence\u001b[0m - AssertionError: A total of 0 metabolites (0.00%) lack annotation with any type\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_sbo.py::\u001b[1mtest_reaction_sbo_presence\u001b[0m - AssertionError: A total of 0 reactions (0.00%) lack annotation with any type\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_sbo.py::\u001b[1mtest_gene_sbo_presence\u001b[0m - AssertionError: A total of 0 genes (0.00%) lack annotation with any type of\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_sbo.py::\u001b[1mtest_metabolic_reaction_specific_sbo_presence\u001b[0m - AssertionError: A total of 0 metabolic reactions (0.00% of all purely\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_sbo.py::\u001b[1mtest_transport_reaction_specific_sbo_presence\u001b[0m - AssertionError: A total of 0 metabolic reactions (0.00% of all transport\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_sbo.py::\u001b[1mtest_metabolite_specific_sbo_presence\u001b[0m - AssertionError: A total of 0 transport reactions (0.00% of all metabolites)\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_sbo.py::\u001b[1mtest_gene_specific_sbo_presence\u001b[0m - AssertionError: A total of 0 genes (0.00% of all genes) lack             an...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_sbo.py::\u001b[1mtest_exchange_specific_sbo_presence\u001b[0m - AssertionError: A total of 0 exchange reactions (0.00% of all exchange\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_sbo.py::\u001b[1mtest_demand_specific_sbo_presence\u001b[0m - AssertionError: A total of 0 genes (0.00% of all demand reactions) lack\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_sbo.py::\u001b[1mtest_sink_specific_sbo_presence\u001b[0m - AssertionError: A total of 0 genes (0.00% of all sink reactions) lack\n",
      "\u001b[31m============ \u001b[31m\u001b[1m87 failed\u001b[0m, \u001b[32m55 passed\u001b[0m, \u001b[33m5 skipped\u001b[0m, \u001b[33m4 warnings\u001b[0m\u001b[31m in 32.21s\u001b[0m\u001b[31m =============\u001b[0m\n",
      "Writing snapshot report to '/mnt/Local_Disk_1/3_MiMB_Book_Chapter/3_MEMOTE/iCN718.html'.\n",
      "The current solver interface glpk doesn't support setting the optimality tolerance.\n",
      "\u001b[1m============================= test session starts ==============================\u001b[0m\n",
      "platform linux -- Python 3.10.16, pytest-8.3.4, pluggy-1.5.0\n",
      "rootdir: /home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests\n",
      "plugins: typeguard-4.4.1, anyio-4.7.0\n",
      "collected 146 items / 1 skipped                                                \u001b[0m\n",
      "\n",
      "../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py \u001b[32m.\u001b[0m\u001b[33m [  0%]\n",
      "\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[31m         [ 44%]\u001b[0m\n",
      "../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_basic.py \u001b[32m.\u001b[0m\u001b[31m [ 45%]\n",
      "\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31m                                                   [ 60%]\u001b[0m\n",
      "../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_biomass.py \u001b[32m.\u001b[0m\u001b[31m [ 60%]\n",
      "\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[31m                                                                [ 67%]\u001b[0m\n",
      "../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py \u001b[32m.\u001b[0m\u001b[31m [ 67%]\n",
      "\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[33ms\u001b[0m\u001b[33ms\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[31m                                             [ 86%]\u001b[0m\n",
      "../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_essentiality.py \u001b[33ms\u001b[0m\u001b[31m [ 87%]\n",
      "\u001b[0m\u001b[31m                                                                         [ 87%]\u001b[0m\n",
      "../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_growth.py \u001b[33ms\u001b[0m\u001b[31m [ 88%]\n",
      "\u001b[0m\u001b[31m                                                                         [ 88%]\u001b[0m\n",
      "../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_matrix.py \u001b[32m.\u001b[0m\u001b[31m [ 89%]\n",
      "\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[32m.\u001b[0m\u001b[31m                                                                      [ 91%]\u001b[0m\n",
      "../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_sbml.py \u001b[32m.\u001b[0m\u001b[31m [ 91%]\n",
      "\u001b[0m\u001b[32m.\u001b[0m\u001b[31m                                                                        [ 92%]\u001b[0m\n",
      "../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_sbo.py \u001b[31mF\u001b[0m\u001b[31m [ 93%]\n",
      "\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[31mF\u001b[0m\u001b[33ms\u001b[0m\u001b[31mF\u001b[0m\u001b[32m.\u001b[0m\u001b[31m                                                               [100%]\u001b[0m\n",
      "\n",
      "\u001b[33m=============================== warnings summary ===============================\u001b[0m\n",
      "../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/_pytest/config/__init__.py:1277\n",
      "  /home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/_pytest/config/__init__.py:1277: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: typeguard\n",
      "    self._mark_plugins_for_rewrite(hook)\n",
      "\n",
      "test_annotation.py::test_metabolite_id_namespace_consistency\n",
      "test_annotation.py::test_reaction_id_namespace_consistency\n",
      "  /home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/support/annotation.py:279: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.\n",
      "  The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.\n",
      "  \n",
      "  For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.\n",
      "  \n",
      "  \n",
      "    df[\"duplicate\"].fillna(False, inplace=True)\n",
      "\n",
      "test_annotation.py::test_metabolite_id_namespace_consistency\n",
      "test_annotation.py::test_reaction_id_namespace_consistency\n",
      "  /home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/support/annotation.py:279: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "    df[\"duplicate\"].fillna(False, inplace=True)\n",
      "\n",
      "-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html\n",
      "\u001b[36m\u001b[1m=========================== short test summary info ============================\u001b[0m\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_overview[pubchem.compound]\u001b[0m - AssertionError: The following 1658 metabolites (100.00%) lack annotation for\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_overview[kegg.compound]\u001b[0m - AssertionError: The following 365 metabolites (22.01%) lack annotation for\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_overview[seed.compound]\u001b[0m - AssertionError: The following 151 metabolites (9.11%) lack annotation for\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_overview[inchikey]\u001b[0m - AssertionError: The following 1658 metabolites (100.00%) lack annotation fo...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_overview[inchi]\u001b[0m - AssertionError: The following 1658 metabolites (100.00%) lack annotation fo...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_overview[chebi]\u001b[0m - AssertionError: The following 284 metabolites (17.13%) lack annotation for ...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_overview[hmdb]\u001b[0m - AssertionError: The following 674 metabolites (40.65%) lack annotation for ...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_overview[reactome]\u001b[0m - AssertionError: The following 1658 metabolites (100.00%) lack annotation fo...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_overview[biocyc]\u001b[0m - AssertionError: The following 266 metabolites (16.04%) lack annotation for ...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_reaction_annotation_overview[rhea]\u001b[0m - AssertionError: The following 1224 reactions (54.11%) lack annotation for r...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_reaction_annotation_overview[kegg.reaction]\u001b[0m - AssertionError: The following 1535 reactions (67.86%) lack annotation for\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_reaction_annotation_overview[seed.reaction]\u001b[0m - AssertionError: The following 515 reactions (22.77%) lack annotation for\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_reaction_annotation_overview[metanetx.reaction]\u001b[0m - AssertionError: The following 9 reactions (0.40%) lack annotation for\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_reaction_annotation_overview[reactome]\u001b[0m - AssertionError: The following 2262 reactions (100.00%) lack annotation for ...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_reaction_annotation_overview[ec-code]\u001b[0m - AssertionError: The following 1200 reactions (53.05%) lack annotation for e...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_reaction_annotation_overview[brenda]\u001b[0m - AssertionError: The following 2262 reactions (100.00%) lack annotation for ...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_reaction_annotation_overview[biocyc]\u001b[0m - AssertionError: The following 1280 reactions (56.59%) lack annotation for b...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_overview[refseq]\u001b[0m - AssertionError: The following 1229 genes (100.00%) lack annotation for refseq:\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_overview[uniprot]\u001b[0m - AssertionError: The following 1229 genes (100.00%) lack annotation for unip...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_overview[ecogene]\u001b[0m - AssertionError: The following 1229 genes (100.00%) lack annotation for ecog...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_overview[kegg.genes]\u001b[0m - AssertionError: The following 1229 genes (100.00%) lack annotation for kegg...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_overview[ncbigi]\u001b[0m - AssertionError: The following 1 genes (0.08%) lack annotation for ncbigi:\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_overview[ncbigene]\u001b[0m - AssertionError: The following 1229 genes (100.00%) lack annotation for ncbi...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_overview[ncbiprotein]\u001b[0m - AssertionError: The following 1229 genes (100.00%) lack annotation for ncbi...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_overview[ccds]\u001b[0m - AssertionError: The following 1229 genes (100.00%) lack annotation for ccds:\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_overview[hprd]\u001b[0m - AssertionError: The following 1229 genes (100.00%) lack annotation for hprd:\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_overview[asap]\u001b[0m - AssertionError: The following 1229 genes (100.00%) lack annotation for asap:\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_wrong_ids[pubchem.compound]\u001b[0m - AssertionError: There are no metabolite annotations for the pubchem.compoun...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_wrong_ids[inchikey]\u001b[0m - AssertionError: There are no metabolite annotations for the inchikey database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_wrong_ids[inchi]\u001b[0m - AssertionError: There are no metabolite annotations for the inchi database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_metabolite_annotation_wrong_ids[reactome]\u001b[0m - AssertionError: There are no metabolite annotations for the reactome database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_reaction_annotation_wrong_ids[rhea]\u001b[0m - AssertionError: A total of 48 reaction annotations (2.12%) do not match the\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_reaction_annotation_wrong_ids[reactome]\u001b[0m - AssertionError: There are no reaction annotations for the reactome database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_reaction_annotation_wrong_ids[ec-code]\u001b[0m - AssertionError: A total of 45 reaction annotations (1.99%) do not match the\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_reaction_annotation_wrong_ids[brenda]\u001b[0m - AssertionError: There are no reaction annotations for the brenda database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_wrong_ids[refseq]\u001b[0m - AssertionError: There are no gene annotations for the refseq database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_wrong_ids[uniprot]\u001b[0m - AssertionError: There are no gene annotations for the uniprot database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_wrong_ids[ecogene]\u001b[0m - AssertionError: There are no gene annotations for the ecogene database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_wrong_ids[kegg.genes]\u001b[0m - AssertionError: There are no gene annotations for the kegg.genes database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_wrong_ids[ncbigi]\u001b[0m - AssertionError: A total of 1228 gene annotations (99.92%) do not match the\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_wrong_ids[ncbigene]\u001b[0m - AssertionError: There are no gene annotations for the ncbigene database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_wrong_ids[ncbiprotein]\u001b[0m - AssertionError: There are no gene annotations for the ncbiprotein database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_wrong_ids[ccds]\u001b[0m - AssertionError: There are no gene annotations for the ccds database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_wrong_ids[hprd]\u001b[0m - AssertionError: There are no gene annotations for the hprd database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_annotation.py::\u001b[1mtest_gene_product_annotation_wrong_ids[asap]\u001b[0m - AssertionError: There are no gene annotations for the asap database.\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_basic.py::\u001b[1mtest_metabolites_formula_presence\u001b[0m - AssertionError: There are a total of 1658         metabolites (100.00%) wit...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_basic.py::\u001b[1mtest_gene_protein_reaction_rule_presence\u001b[0m - AssertionError: There are a total of 77 reactions (3.40%) without GPR:\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_basic.py::\u001b[1mtest_find_reactions_with_partially_identical_annotations\u001b[0m - AssertionError: Based on annotations there are 323 different groups of over...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_basic.py::\u001b[1mtest_find_reactions_with_identical_genes\u001b[0m - AssertionError: Based only on equal genes there are 242 different groups of\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_basic.py::\u001b[1mtest_find_medium_metabolites\u001b[0m - TypeError: unsupported operand type(s) for /: 'int' and 'list'\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_biomass.py::\u001b[1mtest_biomass_consistency[BIOMASS_]\u001b[0m - AssertionError: The component molar mass of the biomass reaction BIOMASS_ s...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_biomass.py::\u001b[1mtest_fast_growth_default[BIOMASS_]\u001b[0m - AssertionError: Using the biomass reaction BIOMASS_ and when the model is s...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_biomass.py::\u001b[1mtest_essential_precursors_not_in_biomass[BIOMASS_]\u001b[0m - AssertionError: BIOMASS_ lacks a total of 8 essential metabolites         (...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_reaction_mass_balance\u001b[0m - AssertionError: A total of 1970 (100.00%) reactions are mass unbalanced wit...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_blocked_reactions\u001b[0m - AssertionError: There are 506 (22.37%) blocked reactions in         the model:\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_find_stoichiometrically_balanced_cycles\u001b[0m - AssertionError: There are 59 (2.61%) reactions         which participate in...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_find_orphans\u001b[0m - AssertionError: A total of 71 (4.28%) metabolites are not produced by any r...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_find_deadends\u001b[0m - AssertionError: A total of 87 (5.25%) metabolites are not consumed by any r...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_find_metabolites_not_produced_with_open_bounds\u001b[0m - AssertionError: A total of 215 (12.97%) metabolites cannot be produced in c...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_consistency.py::\u001b[1mtest_find_metabolites_not_consumed_with_open_bounds\u001b[0m - AssertionError: A total of 428 (25.81%) metabolites cannot be consumed in c...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_sbo.py::\u001b[1mtest_metabolite_sbo_presence\u001b[0m - AssertionError: A total of 0 metabolites (0.00%) lack annotation with any type\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_sbo.py::\u001b[1mtest_reaction_sbo_presence\u001b[0m - AssertionError: A total of 0 reactions (0.00%) lack annotation with any type\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_sbo.py::\u001b[1mtest_gene_sbo_presence\u001b[0m - AssertionError: A total of 0 genes (0.00%) lack annotation with any type of\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_sbo.py::\u001b[1mtest_metabolic_reaction_specific_sbo_presence\u001b[0m - AssertionError: A total of 0 metabolic reactions (0.00% of all purely\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_sbo.py::\u001b[1mtest_transport_reaction_specific_sbo_presence\u001b[0m - AssertionError: A total of 4 metabolic reactions (0.61% of all transport\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_sbo.py::\u001b[1mtest_metabolite_specific_sbo_presence\u001b[0m - AssertionError: A total of 0 transport reactions (0.00% of all metabolites)\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_sbo.py::\u001b[1mtest_gene_specific_sbo_presence\u001b[0m - AssertionError: A total of 0 genes (0.00% of all genes) lack             an...\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_sbo.py::\u001b[1mtest_exchange_specific_sbo_presence\u001b[0m - AssertionError: A total of 0 exchange reactions (0.00% of all exchange\n",
      "\u001b[31mFAILED\u001b[0m ../../../home/pratyay/anaconda3/envs/metquest-protocol/lib/python3.10/site-packages/memote/suite/tests/test_sbo.py::\u001b[1mtest_sink_specific_sbo_presence\u001b[0m - AssertionError: A total of 0 genes (0.00% of all sink reactions) lack\n",
      "\u001b[31m============ \u001b[31m\u001b[1m69 failed\u001b[0m, \u001b[32m72 passed\u001b[0m, \u001b[33m6 skipped\u001b[0m, \u001b[33m5 warnings\u001b[0m\u001b[31m in 47.25s\u001b[0m\u001b[31m =============\u001b[0m\n",
      "Writing snapshot report to '/mnt/Local_Disk_1/3_MiMB_Book_Chapter/3_MEMOTE/iYL1228.html'.\n"
     ]
    }
   ],
   "source": [
    "# Running MEMOTE on the GEMs\n",
    "# Model: iCN718\n",
    "!memote report snapshot --filename \"/mnt/Local_Disk_1/3_MiMB_Book_Chapter/3_MEMOTE/iCN718.html\" /mnt/Local_Disk_1/3_MiMB_Book_Chapter/1_Models/iCN718.xml\n",
    "\n",
    "# Model: iYL1228\n",
    "!memote report snapshot --filename \"/mnt/Local_Disk_1/3_MiMB_Book_Chapter/3_MEMOTE/iYL1228.html\" /mnt/Local_Disk_1/3_MiMB_Book_Chapter/1_Models/iYL1228.xml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def validate_SBML(path_to_models: Union[str, os.PathLike]) -> None:\n",
    "    \"\"\"\n",
    "    Validate SBML (Systems Biology Markup Language) files in a specified directory.\n",
    "\n",
    "    This function reads all `.xml` files in the given directory, checks for errors\n",
    "    in the SBML structure, and logs whether each file is valid or contains errors.\n",
    "\n",
    "    Parameters:\n",
    "        path_to_models (Union[str, os.PathLike]):\n",
    "            The path to the directory containing SBML `.xml` files to validate.\n",
    "\n",
    "    Returns:\n",
    "        None\n",
    "\n",
    "    Raises:\n",
    "        FileNotFoundError: If the specified directory does not exist.\n",
    "        ValueError: If the provided path is not a directory.\n",
    "    \"\"\"\n",
    "    # Ensure the provided path is valid\n",
    "    if not os.path.exists(path_to_models):\n",
    "        raise FileNotFoundError(f\"The path '{path_to_models}' does not exist.\")\n",
    "\n",
    "    if not os.path.isdir(path_to_models):\n",
    "        raise ValueError(f\"The path '{path_to_models}' is not a directory.\")\n",
    "\n",
    "    # Initialize the SBML reader\n",
    "    reader = SBMLReader()\n",
    "\n",
    "    # Iterate through all files in the directory\n",
    "    for file_name in os.listdir(path_to_models):\n",
    "        # Process only `.xml` files\n",
    "        if file_name.endswith('.xml'):\n",
    "            file_path = os.path.join(path_to_models, file_name)\n",
    "\n",
    "            try:\n",
    "                # Read and validate the SBML file\n",
    "                document: SBMLDocument = reader.readSBML(file_path)\n",
    "\n",
    "                if document.getNumErrors() > 0:\n",
    "                    print(f\"[ERROR] SBML file '{file_name}' contains errors!\")\n",
    "                    print(document.getErrorLog().toString())\n",
    "                else:\n",
    "                    print(f\"[SUCCESS] '{file_name}' is a valid SBML file.\")\n",
    "\n",
    "            except Exception as e:\n",
    "                # Log unexpected errors during processing\n",
    "                print(f\"[EXCEPTION] Failed to process '{file_name}': {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[SUCCESS] 'iCN718.xml' is a valid SBML file.\n",
      "[SUCCESS] 'iYL1228.xml' is a valid SBML file.\n"
     ]
    }
   ],
   "source": [
    "# Validating the SBML files\n",
    "validate_SBML(\"/mnt/Local_Disk_1/3_MiMB_Book_Chapter/1_Models/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def min_medium(path_to_models: Union[str, os.PathLike], path_to_medium_file: Union[str, os.PathLike]) -> None:\n",
    "    \"\"\"\n",
    "    Generates the minimal set of metabolites required for multiple organisms to grow\n",
    "    and saves the output to a text file. If the file does not exist, it will be created.\n",
    "\n",
    "    Parameters:\n",
    "        path_to_models (Union[str, os.PathLike]):\n",
    "            Path to the folder containing the model files in SBML format.\n",
    "        path_to_medium_file (Union[str, os.PathLike]):\n",
    "            Path to the output text file where the results will be saved.\n",
    "\n",
    "    Returns:\n",
    "        None\n",
    "\n",
    "    Raises:\n",
    "        FileNotFoundError: If the specified folder for models does not exist.\n",
    "        ValueError: If no valid SBML files are found in the provided folder.\n",
    "    \"\"\"\n",
    "    # Ensure the model folder exists\n",
    "    if not os.path.exists(path_to_models):\n",
    "        raise FileNotFoundError(f\"The path '{path_to_models}' does not exist.\")\n",
    "\n",
    "    if not os.path.isdir(path_to_models):\n",
    "        raise ValueError(f\"The path '{path_to_models}' is not a directory.\")\n",
    "\n",
    "    # Dictionary to store models\n",
    "    models = {}\n",
    "\n",
    "    # Load models from the specified folder\n",
    "    for file_name in os.listdir(path_to_models):\n",
    "        file_path = os.path.join(path_to_models, file_name)\n",
    "        if os.path.isfile(file_path) and file_name.endswith('.xml'):\n",
    "            try:\n",
    "                # Load the model and add it to the dictionary\n",
    "                model = cobra.io.read_sbml_model(file_path)\n",
    "                models[file_name] = model\n",
    "            except Exception as e:\n",
    "                print(f\"[ERROR] Failed to load model '{file_name}': {e}\")\n",
    "\n",
    "    # Raise an error if no models were successfully loaded\n",
    "    if not models:\n",
    "        raise ValueError(\"No valid SBML models were loaded from the specified folder.\")\n",
    "\n",
    "    # List to store the minimal medium metabolites\n",
    "    media: List[str] = []\n",
    "\n",
    "    # Generate minimal medium for each model\n",
    "    for model_name, model in models.items():\n",
    "        try:\n",
    "            # Assume `minimal_medium` is a function that takes a model and returns the minimal medium\n",
    "            minimal = minimal_medium(model, 0.1, minimize_components=True)\n",
    "            media += list(minimal.keys())\n",
    "            print(f\"[SUCCESS] Minimal medium calculated for '{model_name}'.\")\n",
    "        except Exception as e:\n",
    "            print(f\"[ERROR] Failed to calculate minimal medium for '{model_name}': {e}\")\n",
    "\n",
    "    # Remove duplicates by converting to a set and back to a list\n",
    "    media = list(set(media))\n",
    "\n",
    "    # Check if the output file exists and verify write permissions\n",
    "    try:\n",
    "        if os.path.exists(path_to_medium_file):\n",
    "            if not os.access(path_to_medium_file, os.W_OK):\n",
    "                raise PermissionError(f\"The file '{path_to_medium_file}' exists but is not writable.\")\n",
    "        else:\n",
    "            parent_dir = os.path.dirname(path_to_medium_file) or \".\"\n",
    "            if not os.access(parent_dir, os.W_OK):\n",
    "                raise PermissionError(f\"The directory '{parent_dir}' is not writable.\")\n",
    "\n",
    "        # Write the minimal medium metabolites to the specified file\n",
    "        with open(path_to_medium_file, \"w\") as file:\n",
    "            for item in media:\n",
    "                file.write(item + \"\\n\")\n",
    "        print(f\"[SUCCESS] Minimal medium metabolites saved to '{path_to_medium_file}'.\")\n",
    "    except PermissionError as pe:\n",
    "        print(f\"[ERROR] Permission issue: {pe}\")\n",
    "    except Exception as e:\n",
    "        print(f\"[ERROR] Failed to save the minimal medium to '{path_to_medium_file}': {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[SUCCESS] Minimal medium calculated for 'iCN718.xml'.\n",
      "[SUCCESS] Minimal medium calculated for 'iYL1228.xml'.\n",
      "[SUCCESS] Minimal medium metabolites saved to '/mnt/Local_Disk_1/3_MiMB_Book_Chapter/4_Medium/1_Minimal_medium/curated_minimal_medium.txt'.\n"
     ]
    }
   ],
   "source": [
    "# Calculating the minimal medium\n",
    "path_to_models = \"/mnt/Local_Disk_1/3_MiMB_Book_Chapter/1_Models/\" \n",
    "path_to_medium_file = \"/mnt/Local_Disk_1/3_MiMB_Book_Chapter/4_Medium/1_Minimal_medium/curated_minimal_medium.txt\" \n",
    "\n",
    "min_medium(path_to_models, path_to_medium_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_model_info(model_dir: Union[str, os.PathLike]) -> Tuple[Dict[str, str], Dict[str, List[str]], Dict[str, List[str]], List[str], List[str]]:\n",
    "    \"\"\"\n",
    "    Extracts and returns information about SBML models in a given directory.\n",
    "\n",
    "    This function provides the following:\n",
    "    1. `model_id_dict`: A dictionary mapping file names (without .xml extension) to model IDs.\n",
    "    2. `model_compartments_dict`: A dictionary mapping model IDs to a list of compartments present in the model.\n",
    "       Example: 'c' for 'cytosol', 'e' for 'extracellular', 'p' for 'periplasm', 'r' for 'endoplasmic reticulum'.\n",
    "    3. `model_e0_dict`: A dictionary mapping model IDs to lists of reaction IDs containing metabolites ending with \"e0\".\n",
    "    4. `model_id_list`: A list of model IDs.\n",
    "    5. `model_list`: A list of model file names (without .xml extension).\n",
    "\n",
    "    Parameters:\n",
    "        model_dir (Union[str, os.PathLike]):\n",
    "            The path to the directory containing the SBML model files.\n",
    "\n",
    "    Returns:\n",
    "        Tuple containing:\n",
    "            - model_id_dict (Dict[str, str]): File names to model IDs.\n",
    "            - model_compartments_dict (Dict[str, List[str]]): Model IDs to compartment lists.\n",
    "            - model_e0_dict (Dict[str, List[str]]): Model IDs to reaction IDs containing \"e0\" metabolites.\n",
    "            - model_id_list (List[str]): List of model IDs.\n",
    "            - model_list (List[str]): List of file names (without .xml extension).\n",
    "\n",
    "    Raises:\n",
    "        FileNotFoundError: If the directory does not exist.\n",
    "        ValueError: If no valid SBML files are found in the directory.\n",
    "    \"\"\"\n",
    "    if not os.path.exists(model_dir):\n",
    "        raise FileNotFoundError(f\"The directory '{model_dir}' does not exist.\")\n",
    "\n",
    "    if not os.path.isdir(model_dir):\n",
    "        raise ValueError(f\"The path '{model_dir}' is not a directory.\")\n",
    "\n",
    "    model_id_dict = {}\n",
    "    model_compartments_dict = {}\n",
    "    model_e0_dict = {}\n",
    "    model_id_list = []\n",
    "    model_list = []\n",
    "\n",
    "    # Process each file in the directory\n",
    "    for model_file in os.listdir(model_dir):\n",
    "        file_path = os.path.join(model_dir, model_file)\n",
    "        if os.path.isfile(file_path) and model_file.endswith('.xml'):\n",
    "            try:\n",
    "                # Load the SBML model\n",
    "                model = cobra.io.read_sbml_model(file_path)\n",
    "\n",
    "                # Extract model information\n",
    "                file_base_name = model_file.replace(\".xml\", \"\")\n",
    "                model_id_dict[file_base_name] = model.id\n",
    "                model_compartments_dict[model.id] = list(model.compartments.keys())\n",
    "                model_e0_dict[model.id] = []\n",
    "                model_id_list.append(model.id)\n",
    "                model_list.append(file_base_name)\n",
    "\n",
    "                # Identify reactions containing \"e0\" metabolites\n",
    "                for reaction in model.reactions:\n",
    "                    if not reaction.id.startswith(\"EX\"):\n",
    "                        if any(metabolite.id.endswith(\"e0\") for metabolite in reaction.metabolites.keys()):\n",
    "                            model_e0_dict[model.id].append(reaction.id)\n",
    "\n",
    "            except Exception as e:\n",
    "                print(f\"[ERROR] Failed to process model file '{model_file}': {e}\")\n",
    "\n",
    "    # Ensure at least one valid model was processed\n",
    "    if not model_id_dict:\n",
    "        raise ValueError(\"No valid SBML models were found in the directory.\")\n",
    "\n",
    "    return model_id_dict, model_compartments_dict, model_e0_dict, model_id_list, model_list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['iCN718', 'iYL1228']\n",
      "{'iCN718': ['c', 'e', 'p', 'r'], 'iYL1228': ['p', 'c', 'e']}\n"
     ]
    }
   ],
   "source": [
    "# Getting the details of the models\n",
    "model_id_dict, model_compartments_dict, model_e0_dict, model_id_list, model_list = get_model_info(path_to_models)\n",
    "\n",
    "print(model_id_list)\n",
    "print(model_compartments_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_seed(\n",
    "    seed_dir: Union[str, os.PathLike],\n",
    "    site_name: str,\n",
    "    model_compartments_dict: Dict[str, List[str]],\n",
    "    model_id_list: List[str],\n",
    "    path_to_seed_metabolites_file: Union[str, os.PathLike]\n",
    ") -> None:\n",
    "    \"\"\"\n",
    "    Formats the seed metabolites file by adding model IDs and compartments to each metabolite.\n",
    "    Each metabolite is added to every compartment of each model.\n",
    "\n",
    "    Parameters:\n",
    "        seed_dir (Union[str, os.PathLike]):\n",
    "            Path to the folder containing the seed files (minimal medium file and optionally cofactors file).\n",
    "        site_name (str):\n",
    "            The base name of the minimal medium file (without extension).\n",
    "        model_compartments_dict (Dict[str, List[str]]):\n",
    "            A dictionary mapping model IDs to lists of compartments.\n",
    "        model_id_list (List[str]):\n",
    "            A list of model IDs.\n",
    "        path_to_seed_metabolites_file (Union[str, os.PathLike]):\n",
    "            Path to the output file where the formatted seed metabolites will be saved.\n",
    "\n",
    "    Returns:\n",
    "        None\n",
    "\n",
    "    Raises:\n",
    "        FileNotFoundError: If the minimal medium file does not exist.\n",
    "        Exception: If an error occurs while reading or writing files.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        # Read the seed list from the minimal medium file\n",
    "        seed_file_path = os.path.join(seed_dir, site_name + \".txt\")\n",
    "        if not os.path.isfile(seed_file_path):\n",
    "            raise FileNotFoundError(f\"The minimal medium file '{seed_file_path}' does not exist.\")\n",
    "        \n",
    "        with open(seed_file_path, \"r\") as seed_file:\n",
    "            seed_list = seed_file.read().splitlines()\n",
    "\n",
    "        # Read the cofactors list if the file exists\n",
    "        cofactor_file_path = os.path.join(seed_dir, \"cofactors.txt\")\n",
    "        cofactor_list = []\n",
    "        if os.path.isfile(cofactor_file_path):\n",
    "            with open(cofactor_file_path, \"r\") as cofactor_file:\n",
    "                cofactor_list = cofactor_file.read().splitlines()\n",
    "\n",
    "        # Combine seed and cofactor lists into a total set\n",
    "        total: Set[str] = set(seed_list).union(set(cofactor_list))\n",
    "\n",
    "        # Create the formatted seed set\n",
    "        seed_set: Set[str] = set()\n",
    "        for model_id in model_id_list:\n",
    "            compartments = model_compartments_dict.get(model_id, [])\n",
    "            for compartment in compartments:\n",
    "                for metabolite in total:\n",
    "                    formatted_metabolite = f\"{model_id} {metabolite}_{compartment}\"\n",
    "                    seed_set.add(formatted_metabolite)\n",
    "\n",
    "        # Write the formatted seed metabolites to the output file\n",
    "        with open(path_to_seed_metabolites_file, \"w\") as output_file:\n",
    "            for item in sorted(seed_set):\n",
    "                output_file.write(item + \"\\n\")\n",
    "\n",
    "        print(f\"[SUCCESS] Formatted seed metabolites saved to '{path_to_seed_metabolites_file}'.\")\n",
    "\n",
    "    except FileNotFoundError as fnf_error:\n",
    "        print(f\"[ERROR] {fnf_error}\")\n",
    "    except Exception as e:\n",
    "        print(f\"[ERROR] An error occurred: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[SUCCESS] Formatted seed metabolites saved to '/mnt/Local_Disk_1/3_MiMB_Book_Chapter/5_Int_dir/curated_minimal_medium_formatted.txt'.\n"
     ]
    }
   ],
   "source": [
    "# Formatting the seed metabolites\n",
    "seed_dir = \"/mnt/Local_Disk_1/3_MiMB_Book_Chapter/4_Medium/1_Minimal_medium/\"\n",
    "site_name = \"curated_minimal_medium\"\n",
    "path_to_seed_metabolites_file= \"/mnt/Local_Disk_1/3_MiMB_Book_Chapter/5_Int_dir/curated_minimal_medium_formatted.txt\"\n",
    "\n",
    "format_seed(seed_dir,site_name,model_compartments_dict,model_id_list,path_to_seed_metabolites_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Constants for artifact identifiers\n",
    "artifacts = [\n",
    "    \"cpd17041\",  # Biomass\n",
    "    \"cpd17042\",  # DNA replication\n",
    "    \"cpd17043\",  # RNA transcription\n",
    "    \"cpd11416\"   # Protein biosynthesis\n",
    "]\n",
    "\n",
    "def remove_artifacts(\n",
    "    graph: nx.DiGraph,\n",
    "    namemap: Dict[str, List[str]],\n",
    "    model_compartments_dict: Dict[str, List[str]],\n",
    "    model_id_list: List[str] = []\n",
    ") -> nx.DiGraph:\n",
    "    \"\"\"\n",
    "    Removes artifact nodes and associated edges from the graph.\n",
    "\n",
    "    Parameters:\n",
    "        graph (nx.DiGraph):\n",
    "            The directed graph representing the metabolic network.\n",
    "        namemap (Dict[str, List[str]]):\n",
    "            A mapping of node identifiers to names.\n",
    "        model_compartments_dict (Dict[str, List[str]]):\n",
    "            A dictionary mapping model IDs to their compartments.\n",
    "        model_id_list (List[str], optional):\n",
    "            A list of model IDs to filter artifacts.\n",
    "\n",
    "    Returns:\n",
    "        nx.DiGraph: The modified graph with artifact nodes removed.\n",
    "    \"\"\"\n",
    "    global artifacts\n",
    "    remove = set()\n",
    "\n",
    "    # Identify nodes with 'bio1' in their names for removal\n",
    "    for key, values in namemap.items():\n",
    "        if \"bio1\" in values:\n",
    "            remove.add(key)\n",
    "\n",
    "    # Create a set of artifact-related nodes\n",
    "    artifacts_set = set()\n",
    "    for artifact in artifacts:\n",
    "        for model_id in model_id_list:\n",
    "            for compartment in model_compartments_dict.get(model_id, []):\n",
    "                node = f\"{artifact}_{compartment}\"\n",
    "                if graph.has_node(node):\n",
    "                    artifacts_set.add(node)\n",
    "                model_node = f\"{model_id}_{compartment}\"\n",
    "                if graph.has_node(model_node):\n",
    "                    artifacts_set.add(f\"{model_id} {node}\")\n",
    "\n",
    "    # Add edges connected to artifact nodes for removal\n",
    "    for artifact_node in artifacts_set:\n",
    "        for in_edge in graph.out_edges(artifact_node):\n",
    "            remove.update(in_edge)\n",
    "        for out_edge in graph.in_edges(artifact_node):\n",
    "            remove.update(out_edge)\n",
    "\n",
    "    # Remove identified nodes\n",
    "    for node in remove:\n",
    "        if graph.has_node(node):\n",
    "            graph.remove_node(node)\n",
    "\n",
    "    return graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_individual_graph(\n",
    "    model_dir: str, \n",
    "    model_id_dict: Dict[str, str], \n",
    "    model_compartments_dict: Dict[str, str], \n",
    "    model_list: list,\n",
    "    intermediate_dir: str\n",
    ") -> Tuple[Dict[str, nx.Graph], Dict[str, Dict]]:\n",
    "    \"\"\"\n",
    "    Create individual graphs for models and save them for later use.\n",
    "\n",
    "    Parameters:\n",
    "    ----------\n",
    "    model_dir : str\n",
    "        Path to the folder containing the models.\n",
    "    model_id_dict : Dict[str, str]\n",
    "        Dictionary mapping model names to their IDs.\n",
    "    model_compartments_dict : Dict[str, str]\n",
    "        Dictionary containing compartment details for the models.\n",
    "    model_list : list\n",
    "        List of model names to process.\n",
    "    intermediate_dir : str\n",
    "        Path to the directory to save intermediate files.\n",
    "\n",
    "    Returns:\n",
    "    -------\n",
    "    None\n",
    "    \"\"\"\n",
    "    print(\"[INFO] Creating individual graphs...\")\n",
    "\n",
    "    individual_graph_dict = {}\n",
    "    individual_namemap_dict = {}\n",
    "\n",
    "    os.makedirs(intermediate_dir, exist_ok=True)\n",
    "\n",
    "    for model_name in model_list:\n",
    "        file_path = os.path.join(model_dir, f\"{model_name}.xml\")\n",
    "        if file_path.endswith(\".xml\") and os.path.isfile(file_path):\n",
    "            # Copy the model file to intermediate directory\n",
    "            intermediate_file = os.path.join(intermediate_dir, f\"{model_name}.xml\")\n",
    "            shutil.copyfile(file_path, intermediate_file)\n",
    "\n",
    "    # Create graph\n",
    "    num = 1\n",
    "    graph, namemap = create_graph(intermediate_dir, num)\n",
    "    graph = remove_artifacts(\n",
    "        graph, namemap, model_compartments_dict, [model_id_dict[model_name]]\n",
    "    )\n",
    "    \n",
    "    print(f\"[INFO] Saved graph and namemap for all the models in {intermediate_dir}\")\n",
    "\n",
    "    return individual_graph_dict, individual_namemap_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[INFO] Creating individual graphs...\n",
      "Filenames ['iCN718.xml', 'iYL1228.xml']\n",
      "Number of edges in graph 6194\n",
      "Number of nodes in graph 2467\n",
      "Graph and namemap saved for file(s) in /mnt/Local_Disk_1/3_MiMB_Book_Chapter/5_Int_dir/\n",
      "Number of edges in graph 11214\n",
      "Number of nodes in graph 5039\n",
      "Graph and namemap saved for file(s) in /mnt/Local_Disk_1/3_MiMB_Book_Chapter/5_Int_dir/\n",
      "[INFO] Saved graph and namemap for all the models in /mnt/Local_Disk_1/3_MiMB_Book_Chapter/5_Int_dir/\n"
     ]
    }
   ],
   "source": [
    "# Getting the individual graphs\n",
    "model_dir = \"/mnt/Local_Disk_1/3_MiMB_Book_Chapter/1_Models/\"\n",
    "intermediate_dir = \"/mnt/Local_Disk_1/3_MiMB_Book_Chapter/5_Int_dir/\"\n",
    "\n",
    "individual_graph_dict, individual_namemap_dict = get_individual_graph(model_dir, model_id_dict, model_compartments_dict, model_list, intermediate_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pair_graph(\n",
    "    model_dir: str, \n",
    "    model_id_dict: Dict[str, str], \n",
    "    model_compartments_dict: Dict[str, str], \n",
    "    model_list: list,\n",
    "    intermediate_dir: str\n",
    ") -> Tuple[Dict[str, nx.Graph], Dict[str, Dict]]:\n",
    "    \"\"\"\n",
    "    Create pairwise graphs for combinations of models and save them for later use.\n",
    "\n",
    "    Parameters:\n",
    "    ----------\n",
    "    model_dir : str\n",
    "        Path to the folder containing the models.\n",
    "    model_id_dict : Dict[str, str]\n",
    "        Dictionary mapping model names to their IDs.\n",
    "    model_compartments_dict : Dict[str, str]\n",
    "        Dictionary containing compartment details for the models.\n",
    "    model_list : list\n",
    "        List of model names to process.\n",
    "    intermediate_dir : str\n",
    "        Path to the directory to save intermediate files.\n",
    "\n",
    "    Returns:\n",
    "    -------\n",
    "    pair_graph_dict : Dict[str, nx.Graph]\n",
    "        Dictionary of pairwise graphs for each combination of models.\n",
    "    pair_namemap_dict : Dict[str, Dict]\n",
    "        Dictionary of name maps for each combination of models.\n",
    "    \"\"\"\n",
    "    print(\"[INFO] Creating pairwise graphs...\")\n",
    "\n",
    "    pair_graph_dict = {}\n",
    "    pair_namemap_dict = {}\n",
    "    done = []\n",
    "\n",
    "    os.makedirs(intermediate_dir, exist_ok=True)\n",
    "\n",
    "    combinations_list = combinations(model_list, 2)\n",
    "    for comb in combinations_list:\n",
    "        pair_name = f\"{comb[0]}_{comb[1]}\"\n",
    "        if pair_name not in done and f\"{comb[1]}_{comb[0]}\" not in done:\n",
    "            done.append(pair_name)\n",
    "\n",
    "            file1 = os.path.join(model_dir, f\"{comb[0]}.xml\")\n",
    "            file2 = os.path.join(model_dir, f\"{comb[1]}.xml\")\n",
    "\n",
    "            if (file1.endswith(\".xml\") and os.path.isfile(file1) and\n",
    "                file2.endswith(\".xml\") and os.path.isfile(file2)):\n",
    "\n",
    "                # Copy the model files to intermediate directory\n",
    "                intermediate_file1 = os.path.join(intermediate_dir, f\"{comb[0]}.xml\")\n",
    "                intermediate_file2 = os.path.join(intermediate_dir, f\"{comb[1]}.xml\")\n",
    "\n",
    "                shutil.copyfile(file1, intermediate_file1)\n",
    "                shutil.copyfile(file2, intermediate_file2)\n",
    "\n",
    "                # Create pairwise graph\n",
    "                num = 2\n",
    "                graph, namemap = create_graph(intermediate_dir, num)\n",
    "                graph = remove_artifacts(\n",
    "                    graph, namemap, model_compartments_dict, [\n",
    "                        model_id_dict[comb[0]], model_id_dict[comb[1]]\n",
    "                    ]\n",
    "                )\n",
    "\n",
    "                # Save the graph and namemap\n",
    "                graph_pickle_path = os.path.join(intermediate_dir, f\"{pair_name}_graph.pickle\")\n",
    "                namemap_pickle_path = os.path.join(intermediate_dir, f\"{pair_name}_namemap.pickle\")\n",
    "\n",
    "                with open(graph_pickle_path, 'wb') as graph_file:\n",
    "                    pickle.dump(graph, graph_file, pickle.HIGHEST_PROTOCOL)\n",
    "                with open(namemap_pickle_path, 'wb') as namemap_file:\n",
    "                    pickle.dump(namemap, namemap_file, pickle.HIGHEST_PROTOCOL)\n",
    "\n",
    "                pair_graph_dict[pair_name] = graph\n",
    "                pair_namemap_dict[pair_name] = namemap\n",
    "\n",
    "                print(f\"[INFO] Saved graph and namemap for pair: {pair_name}\")\n",
    "\n",
    "    return pair_graph_dict, pair_namemap_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[INFO] Creating pairwise graphs...\n",
      "Filenames ['iCN718.xml', 'iYL1228.xml']\n",
      "Number of edges in graph 17408\n",
      "Number of nodes in graph 7400\n",
      "Graph and namemap saved for file(s) in /mnt/Local_Disk_1/3_MiMB_Book_Chapter/5_Int_dir/\n",
      "[INFO] Saved graph and namemap for pair: iCN718_iYL1228\n"
     ]
    }
   ],
   "source": [
    "# Getting the pair graphs\n",
    "model_dir = \"/mnt/Local_Disk_1/3_MiMB_Book_Chapter/1_Models/\"\n",
    "intermediate_dir = \"/mnt/Local_Disk_1/3_MiMB_Book_Chapter/5_Int_dir/\"\n",
    "\n",
    "pair_graph_dict, pair_namemap_dict = get_pair_graph(model_dir, model_id_dict, model_compartments_dict, model_list, intermediate_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_metabolic_support_index(intermediate_dir, graph_name_combined, graph_name_firstorg, graph_name_secondorg, seed_metabolite_fname):\n",
    "    \"\"\"\n",
    "    Calculate the Metabolic Support Index (MSI), which defines the fraction of blocked\n",
    "    or stuck reactions relieved by the presence of another organism. This index is useful\n",
    "    for assessing the metabolic support between two organisms within a combined metabolic network.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    intermediate_dir : str\n",
    "        Path to the directory containing all data files (e.g., graphs, metabolites).\n",
    "    graph_name_combined : str\n",
    "        Filename of the combined graph (joint metabolic network) between two organisms.\n",
    "    graph_name_firstorg : str\n",
    "        Filename of the first organism's graph (metabolic network).\n",
    "    graph_name_secondorg : str\n",
    "        Filename of the second organism's graph (metabolic network).\n",
    "    seed_metabolite_fname : str\n",
    "        Filename containing the seed metabolite list that will be used to perform the forward pass.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    None\n",
    "        Outputs MSI values to a CSV file and prints results to the console. The MSI indicates the fraction\n",
    "        of metabolic reactions that are relieved (stuck reactions are \"unblocked\") when the two organisms\n",
    "        are in the same environment.\n",
    "    \"\"\"\n",
    "\n",
    "    # Helper function to load a graph from a pickle file\n",
    "    def load_pickle_graph(filepath):\n",
    "        \"\"\"\n",
    "        Load a metabolic network graph from a pickle file.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        filepath : str\n",
    "            Path to the pickle file containing the graph.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        networkx.Graph\n",
    "            Loaded graph object.\n",
    "        \"\"\"\n",
    "        with open(filepath, \"rb\") as file:\n",
    "            return pickle.load(file)\n",
    "\n",
    "    print(\"[INFO] Loading graph files...\")\n",
    "\n",
    "    # Load the combined and individual organism graphs\n",
    "    combined_graph = load_pickle_graph(os.path.join(intermediate_dir, graph_name_combined))\n",
    "    first_org_graph = load_pickle_graph(os.path.join(intermediate_dir, graph_name_firstorg))\n",
    "    second_org_graph = load_pickle_graph(os.path.join(intermediate_dir, graph_name_secondorg))\n",
    "\n",
    "    print(\"[INFO] Loading seed metabolite file...\")\n",
    "\n",
    "    # Load seed metabolites from the provided file\n",
    "    with open(os.path.join(intermediate_dir, seed_metabolite_fname), \"r\") as f:\n",
    "        seed_metabolites = set(f.read().splitlines())\n",
    "\n",
    "    # Run MetQuest's forward pass to calculate metabolic scope for all graphs\n",
    "    print(\"[INFO] Performing metabolic scope calculations for combined and individual graphs...\")\n",
    "    lb_m, status_dict, scope = metquest.forward_pass(combined_graph, seed_metabolites)\n",
    "    lb_m_org_1, status_dict_org_1, scope_org_1 = metquest.forward_pass(first_org_graph, seed_metabolites)\n",
    "    lb_m_org_2, status_dict_org_2, scope_org_2 = metquest.forward_pass(second_org_graph, seed_metabolites)\n",
    "\n",
    "    # Extract reactions and metabolites for both organisms\n",
    "    def extract_reactions_and_metabolites(graph):\n",
    "        \"\"\"\n",
    "        Extract reaction and metabolite nodes from a metabolic network graph.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        graph : networkx.Graph\n",
    "            The metabolic network graph from which reactions and metabolites are to be extracted.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        reactions : list\n",
    "            List of reaction nodes.\n",
    "        metabolites : list\n",
    "            List of metabolite nodes.\n",
    "        \"\"\"\n",
    "        reactions, metabolites = [], []\n",
    "        node_attr = nx.get_node_attributes(graph, 'bipartite')\n",
    "        for node in graph:\n",
    "            if node_attr[node] == 1:  # Reaction nodes\n",
    "                reactions.append(node)\n",
    "            else:  # Metabolite nodes\n",
    "                met_name = node.split(' ')[0]\n",
    "                if '_e' not in met_name:  # Exclude extracellular metabolites\n",
    "                    metabolites.append(met_name)\n",
    "        return reactions, metabolites\n",
    "\n",
    "    # Extract reactions and metabolites for both organisms\n",
    "    org_1_reactions, org_1_metabolites = extract_reactions_and_metabolites(first_org_graph)\n",
    "    org_2_reactions, org_2_metabolites = extract_reactions_and_metabolites(second_org_graph)\n",
    "\n",
    "    # Identify organism labels based on metabolite names (first metabolite name)\n",
    "    org_1 = org_1_metabolites[0]\n",
    "    org_2 = org_2_metabolites[0]\n",
    "\n",
    "    print(f\"[INFO] Organism 1 identified as: {org_1}\")\n",
    "    print(f\"[INFO] Organism 2 identified as: {org_2}\")\n",
    "\n",
    "    # Identify reactions in the combined graph related to each organism\n",
    "    reactions = []\n",
    "    org_1_reactions_combined = []\n",
    "    org_2_reactions_combined = []\n",
    "\n",
    "    node_attr = nx.get_node_attributes(combined_graph, 'bipartite')\n",
    "    for node in combined_graph:\n",
    "        if node_attr[node] == 1:  # Reaction node\n",
    "            reactions.append(node)\n",
    "            if org_1 in node:\n",
    "                org_1_reactions_combined.append(node)\n",
    "            if org_2 in node:\n",
    "                org_2_reactions_combined.append(node)\n",
    "\n",
    "    def classify_reactions(reactions_combined):\n",
    "        \"\"\"\n",
    "        Classify reactions as either internal or external based on specific tags.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        reactions_combined : list\n",
    "            List of reaction nodes from the combined graph.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        internal : list\n",
    "            List of internal reactions.\n",
    "        external : list\n",
    "            List of external reactions.\n",
    "        \"\"\"\n",
    "        internal, external = [], []\n",
    "        for reaction in reactions_combined:\n",
    "            if 'ER' not in reaction:  # 'ER' tag for external reactions\n",
    "                internal.append(reaction)\n",
    "            else:\n",
    "                external.append(reaction)\n",
    "        return internal, external\n",
    "\n",
    "    # Classify reactions for both organisms in the combined graph\n",
    "    org_1_internal, org_1_external = classify_reactions(org_1_reactions_combined)\n",
    "    org_2_internal, org_2_external = classify_reactions(org_2_reactions_combined)\n",
    "\n",
    "    # Calculate stuck reactions in the combined graph (those not active in the forward pass)\n",
    "    all_stuck_nodes = set(reactions) - set(status_dict)\n",
    "\n",
    "    def classify_stuck_reactions(stuck_nodes, organism_label):\n",
    "        \"\"\"\n",
    "        Classify stuck reactions as internal or external based on the organism label.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        stuck_nodes : set\n",
    "            Set of stuck reaction nodes.\n",
    "        organism_label : str\n",
    "            Organism label used to filter reactions related to the organism.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        internal : list\n",
    "            List of internal stuck reactions for the given organism.\n",
    "        external : set\n",
    "            Set of external stuck reactions for the given organism.\n",
    "        \"\"\"\n",
    "        internal, external = [], set()\n",
    "        for node in stuck_nodes:\n",
    "            if organism_label in node:\n",
    "                if 'NCE' not in node and 'ER' not in node:\n",
    "                    if any(tag in node for tag in ['IR', 'RR', 'RevBR']):  # Internal reaction tags\n",
    "                        internal.append(node)\n",
    "                else:\n",
    "                    external.add(node)\n",
    "        return internal, external\n",
    "\n",
    "    # Classify stuck reactions for both organisms\n",
    "    stuck_org_1_internal, stuck_org_1_external = classify_stuck_reactions(all_stuck_nodes, org_1)\n",
    "    stuck_org_2_internal, stuck_org_2_external = classify_stuck_reactions(all_stuck_nodes, org_2)\n",
    "\n",
    "    # Calculate individual stuck reactions for each organism\n",
    "    def identify_stuck_reactions(status_dict, reactions):\n",
    "        \"\"\"\n",
    "        Identify the stuck reactions for a given organism based on the status dictionary.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        status_dict : dict\n",
    "            The status dictionary from MetQuest's forward pass.\n",
    "        reactions : list\n",
    "            List of reaction nodes for the organism.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        internal : list\n",
    "            List of internal stuck reactions for the organism.\n",
    "        external : list\n",
    "            List of external stuck reactions for the organism.\n",
    "        \"\"\"\n",
    "        stuck_nodes = set(reactions) - set(status_dict)\n",
    "        internal, external = [], []\n",
    "        for node in stuck_nodes:\n",
    "            if any(tag in node for tag in ['IR', 'RR', 'RevBR']):  # Internal reaction tags\n",
    "                if 'E' not in node:  # Exclude external reactions\n",
    "                    internal.append(node)\n",
    "                else:\n",
    "                    external.append(node)\n",
    "            else:\n",
    "                external.append(node)\n",
    "        return internal, external\n",
    "\n",
    "    # Identify individual stuck reactions for each organism\n",
    "    org_1_stuck_internal, org_1_stuck_external = identify_stuck_reactions(status_dict_org_1, org_1_reactions)\n",
    "    org_2_stuck_internal, org_2_stuck_external = identify_stuck_reactions(status_dict_org_2, org_2_reactions)\n",
    "\n",
    "    # Calculate MSI values for both organisms based on internal stuck reactions\n",
    "    def calculate_msi(stuck_individual, stuck_combined):\n",
    "        \"\"\"\n",
    "        Calculate the Metabolic Support Index (MSI) as the fraction of stuck reactions relieved\n",
    "        by the combined metabolic environment.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        stuck_individual : list\n",
    "            List of individual stuck reactions.\n",
    "        stuck_combined : list\n",
    "            List of combined stuck reactions (when both organisms are present).\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        float\n",
    "            The calculated MSI value.\n",
    "        \"\"\"\n",
    "        return (len(stuck_individual) - len(stuck_combined)) / len(stuck_individual)\n",
    "\n",
    "    # Calculate MSI for both organisms\n",
    "    msi_org_1 = calculate_msi(org_1_stuck_internal, stuck_org_1_internal)\n",
    "    msi_org_2 = calculate_msi(org_2_stuck_internal, stuck_org_2_internal)\n",
    "\n",
    "    # Print the calculated MSI values\n",
    "    print(\"[INFO] Calculating Metabolic Support Index (MSI)...\")\n",
    "    print(f\"[INFO] MSI ({org_1} in presence of {org_2}): {msi_org_1:.4f}\")\n",
    "    print(f\"[INFO] MSI ({org_2} in presence of {org_1}): {msi_org_2:.4f}\")\n",
    "\n",
    "    # Output MSI to CSV\n",
    "    output_dir = os.path.join(intermediate_dir, 'results')\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "    output_file = org_1 + '_' + org_2 + '_' + seed_metabolite_fname[:-14] + '_MSI.csv'\n",
    "\n",
    "    msi_output_file = os.path.join(output_dir, output_file)\n",
    "    with open(msi_output_file, 'w') as f:\n",
    "        f.write('Organism,In_the_presence_of,MSI_value\\n')\n",
    "        f.write(f\"{org_1},{org_2},{msi_org_1:.4f}\\n\")\n",
    "        f.write(f\"{org_2},{org_1},{msi_org_2:.4f}\\n\")\n",
    "\n",
    "    print(f\"[SUCCESS] MSI values have been written to: {msi_output_file}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[INFO] Loading graph files...\n",
      "[INFO] Loading seed metabolite file...\n",
      "[INFO] Performing metabolic scope calculations for combined and individual graphs...\n",
      "[INFO] Organism 1 identified as: iCN718\n",
      "[INFO] Organism 2 identified as: iYL1228\n",
      "[INFO] Calculating Metabolic Support Index (MSI)...\n",
      "[INFO] MSI (iCN718 in presence of iYL1228): 0.0998\n",
      "[INFO] MSI (iYL1228 in presence of iCN718): 0.0076\n",
      "[SUCCESS] MSI values have been written to: /mnt/Local_Disk_1/3_MiMB_Book_Chapter/5_Int_dir/results/iCN718_iYL1228_curated_minimal_medium_MSI.csv\n"
     ]
    }
   ],
   "source": [
    "# Calculating the Metabolic Support Index (MSI)\n",
    "graph_name_combined = \"iCN718_iYL1228_.gpickle\"\n",
    "graph_name_firstorg = \"iCN718_.gpickle\"\n",
    "graph_name_secondorg = \"iYL1228_.gpickle\"\n",
    "seed_metabolite_fname = \"curated_minimal_medium_formatted.txt\"\n",
    "\n",
    "find_metabolic_support_index(intermediate_dir, graph_name_combined, graph_name_firstorg, graph_name_secondorg, seed_metabolite_fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[SUCCESS] Formatted seed metabolites saved to '/mnt/Local_Disk_1/3_MiMB_Book_Chapter/5_Int_dir/M9_medium_formatted.txt'.\n",
      "[SUCCESS] Formatted seed metabolites saved to '/mnt/Local_Disk_1/3_MiMB_Book_Chapter/5_Int_dir/LB_aerobic_formatted.txt'.\n",
      "[SUCCESS] Formatted seed metabolites saved to '/mnt/Local_Disk_1/3_MiMB_Book_Chapter/5_Int_dir/LB_anaerobic_formatted.txt'.\n"
     ]
    }
   ],
   "source": [
    "# Iteratively for other medium files\n",
    "# Formatting M9 medium\n",
    "seed_dir = \"/mnt/Local_Disk_1/3_MiMB_Book_Chapter/4_Medium/2_M9_medium/\"\n",
    "site_name = \"M9_medium\"\n",
    "path_to_seed_metabolites_file= \"/mnt/Local_Disk_1/3_MiMB_Book_Chapter/5_Int_dir/M9_medium_formatted.txt\"\n",
    "\n",
    "format_seed(seed_dir,site_name,model_compartments_dict,model_id_list,path_to_seed_metabolites_file)\n",
    "\n",
    "# Formatting LB medium (aerobic)\n",
    "seed_dir = \"/mnt/Local_Disk_1/3_MiMB_Book_Chapter/4_Medium/\"\n",
    "site_name = \"LB_aerobic\"\n",
    "path_to_seed_metabolites_file= \"/mnt/Local_Disk_1/3_MiMB_Book_Chapter/5_Int_dir/LB_aerobic_formatted.txt\"\n",
    "\n",
    "format_seed(seed_dir,site_name,model_compartments_dict,model_id_list,path_to_seed_metabolites_file)\n",
    "\n",
    "# Formatting LB medium (anaerobic)\n",
    "seed_dir = \"/mnt/Local_Disk_1/3_MiMB_Book_Chapter/4_Medium/\"\n",
    "site_name = \"LB_anaerobic\"\n",
    "path_to_seed_metabolites_file= \"/mnt/Local_Disk_1/3_MiMB_Book_Chapter/5_Int_dir/LB_anaerobic_formatted.txt\"\n",
    "\n",
    "format_seed(seed_dir,site_name,model_compartments_dict,model_id_list,path_to_seed_metabolites_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[INFO] Loading graph files...\n",
      "[INFO] Loading seed metabolite file...\n",
      "[INFO] Performing metabolic scope calculations for combined and individual graphs...\n",
      "[INFO] Organism 1 identified as: iCN718\n",
      "[INFO] Organism 2 identified as: iYL1228\n",
      "[INFO] Calculating Metabolic Support Index (MSI)...\n",
      "[INFO] MSI (iCN718 in presence of iYL1228): 0.1101\n",
      "[INFO] MSI (iYL1228 in presence of iCN718): 0.0079\n",
      "[SUCCESS] MSI values have been written to: /mnt/Local_Disk_1/3_MiMB_Book_Chapter/5_Int_dir/results/iCN718_iYL1228_M9_medium_MSI.csv\n",
      "[INFO] Loading graph files...\n",
      "[INFO] Loading seed metabolite file...\n",
      "[INFO] Performing metabolic scope calculations for combined and individual graphs...\n",
      "[INFO] Organism 1 identified as: iCN718\n",
      "[INFO] Organism 2 identified as: iYL1228\n",
      "[INFO] Calculating Metabolic Support Index (MSI)...\n",
      "[INFO] MSI (iCN718 in presence of iYL1228): 0.0110\n",
      "[INFO] MSI (iYL1228 in presence of iCN718): 0.0019\n",
      "[SUCCESS] MSI values have been written to: /mnt/Local_Disk_1/3_MiMB_Book_Chapter/5_Int_dir/results/iCN718_iYL1228_LB_aerobic_MSI.csv\n",
      "[INFO] Loading graph files...\n",
      "[INFO] Loading seed metabolite file...\n",
      "[INFO] Performing metabolic scope calculations for combined and individual graphs...\n",
      "[INFO] Organism 1 identified as: iCN718\n",
      "[INFO] Organism 2 identified as: iYL1228\n",
      "[INFO] Calculating Metabolic Support Index (MSI)...\n",
      "[INFO] MSI (iCN718 in presence of iYL1228): 0.0288\n",
      "[INFO] MSI (iYL1228 in presence of iCN718): 0.0019\n",
      "[SUCCESS] MSI values have been written to: /mnt/Local_Disk_1/3_MiMB_Book_Chapter/5_Int_dir/results/iCN718_iYL1228_LB_anaerobic_MSI.csv\n"
     ]
    }
   ],
   "source": [
    "# Calculating the Metabolic Support Index (MSI) for M9 medium\n",
    "seed_metabolite_fname = \"M9_medium_formatted.txt\"\n",
    "\n",
    "find_metabolic_support_index(intermediate_dir, graph_name_combined, graph_name_firstorg, graph_name_secondorg, seed_metabolite_fname)\n",
    "\n",
    "# Calculating the Metabolic Support Index (MSI) for LB medium (aerobic)\n",
    "seed_metabolite_fname = \"LB_aerobic_formatted.txt\"\n",
    "\n",
    "find_metabolic_support_index(intermediate_dir, graph_name_combined, graph_name_firstorg, graph_name_secondorg, seed_metabolite_fname)\n",
    "\n",
    "# Calculating the Metabolic Support Index (MSI) for LB medium (anaerobic)\n",
    "seed_metabolite_fname = \"LB_anaerobic_formatted.txt\"\n",
    "\n",
    "find_metabolic_support_index(intermediate_dir, graph_name_combined, graph_name_firstorg, graph_name_secondorg, seed_metabolite_fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "metquest-protocol",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
